store/statstore/stat_store.go

```diff
@@ -0,0 +1,486 @@
+	ss.validCache = false
+
+	// Determine how many resolutions in the future the new point is at.
+	// The StatStore always represents values up until 1 resolution from lastPut.
+	// If the TimePoint is more than one resolutions in the future, the last bucket is -
+	// extended to be exactly one resolution behind the new lastPut timestamp.
+	numRes := uint8(0)
+	curr := ts
+	for curr.After(ss.lastPut.stamp) {
+		curr = curr.Add(-ss.resolution)
+		numRes++
+	}
+
+	// Create a new bucket if the buffer is empty
+	if ss.buffer.Front() == nil {
+		ss.newBucket(numRes)
+		ss.resetLastPut(ts, tp.Value)
+		for ss.tpCount > ss.windowDuration {
+			ss.rewind()
+		}
+		return nil
+	}
+
+	lastEntry := ss.buffer.Front().Value.(tpBucket)
+	lastAvg := ss.lastPut.average
+
+	// Place lastPut in the latest bucket if the difference from its average
+	// is less than epsilon
+	if uint64(math.Abs(float64(lastEntry.value)-lastAvg)) < ss.epsilon {
+		lastEntry.count += numRes
+		ss.tpCount += numRes
+		if ss.lastPut.max > lastEntry.max {
+			lastEntry.max = ss.lastPut.max
+			lastEntry.maxIdx = lastEntry.count - 1
+		}
+		// update in list
+		ss.buffer.Front().Value = lastEntry
+	} else {
+		// Create a new bucket
+		ss.newBucket(numRes)
+	}
+
+	// Delete the earliest represented TimePoints if the window is full
+	for ss.tpCount > ss.windowDuration {
+		ss.rewind()
+	}
+
+	ss.resetLastPut(ts, tp.Value)
+	return nil
+}
+
+// resetLastPut initializes the lastPut field of the StatStore, given a time and a value.
+func (ss *StatStore) resetLastPut(timestamp time.Time, value uint64) {
+	ss.lastPut.stamp = timestamp
+	ss.lastPut.actualCount = 1
+	ss.lastPut.average = float64(value)
+	ss.lastPut.max = value
+}
+
+// newBucket appends a new bucket to the StatStore, using the values of lastPut.
+// newBucket should be always called BEFORE resetting lastPut.
+// newBuckets are created by rounding up the lastPut average to the closest epsilon.
+// numRes represents the number of resolutions from the newest TimePoint to the lastPut.
+// numRes resolutions will be represented in the newly created bucket.
+func (ss *StatStore) newBucket(numRes uint8) {
+	newEntry := tpBucket{
+		count:  numRes,
+		value:  ((uint64(ss.lastPut.average) / ss.epsilon) + 1) * ss.epsilon,
+		max:    ss.lastPut.max,
+		maxIdx: 0,
+	}
+	ss.buffer.PushFront(newEntry)
+	ss.tpCount += numRes
+
+	// If this was the first bucket, update ss.start
+	if ss.start.Equal(time.Time{}) {
+		ss.start = ss.lastPut.stamp
+	}
+}
+
+// rewind deletes the oldest one resolution of data in the StatStore.
+func (ss *StatStore) rewind() {
+	firstElem := ss.buffer.Back()
+	firstEntry := firstElem.Value.(tpBucket)
+	// Decrement number of TimePoints in the earliest tpBucket
+	firstEntry.count--
+	// Decrement total number of TimePoints in the StatStore
+	ss.tpCount--
+	if firstEntry.maxIdx == 0 {
+		// The Max value was just removed, lose precision for other maxes in this bucket
+		firstEntry.max = firstEntry.value
```

Comment:
this is the bucket max, and we cannot recalculate it as we have lost
information of the other maxes in that bucket
On Aug 12, 2015 12:14 PM, "Vish Kannan" <notifications@github.com> wrote:

> In store/statstore/stat_store.go
> <https://github.com/kubernetes/heapster/pull/461#discussion_r36901283>:
>
> > +	if ss.start.Equal(time.Time{}) {
> > +		ss.start = ss.lastPut.stamp
> > +	}
> > +}
> > +
> > +// rewind deletes the oldest one resolution of data in the StatStore.
> > +func (ss *StatStore) rewind() {
> > +	firstElem := ss.buffer.Back()
> > +	firstEntry := firstElem.Value.(tpBucket)
> > +	// Decrement number of TimePoints in the earliest tpBucket
> > +	firstEntry.count--
> > +	// Decrement total number of TimePoints in the StatStore
> > +	ss.tpCount--
> > +	if firstEntry.maxIdx == 0 {
> > +		// The Max value was just removed, lose precision for other maxes in this bucket
> > +		firstEntry.max = firstEntry.value
>
> Why not calculate the new max?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/heapster/pull/461/files#r36901283>.
>